src/processor/operator/recursive_extend.cpp

```diff
@@ -77,38 +80,40 @@ void RecursiveExtend::executeInternal(ExecutionContext* context) {
             totalNumNodes += graph->getMaxOffset(clientContext->getTransaction(), tableID);
         }
     }
+    std::string propertyName;
+    if (bindData.weightPropertyExpr != nullptr) {
+        propertyName =
+            bindData.weightPropertyExpr->ptrCast<PropertyExpression>()->getPropertyName();
+    }
     offset_t completedNumNodes = 0;
     for (auto& tableID : graph->getNodeTableIDs()) {
         if (!inputNodeMaskMap->containsTableID(tableID)) {
             continue;
         }
-        auto calcFunc = [tableID, graph, context, clientContext, this](offset_t offset) {
+        auto calcFunc = [tableID, propertyName, graph, context, this](offset_t offset) {
+            auto clientContext = context->clientContext;
             if (clientContext->interrupted()) {
                 throw InterruptException{};
             }
+            auto computeState = function->getComputeState(context, bindData, sharedState.get());
             auto sourceNodeID = nodeID_t{offset, tableID};
-            auto rjCompState =
-                function->getRJCompState(context, sourceNodeID, bindData, sharedState.get());
-            auto gdsComputeState = rjCompState.gdsComputeState.get();
-            gdsComputeState->initSource(sourceNodeID);
-            std::string propertyName;
-            if (bindData.weightPropertyExpr != nullptr) {
-                propertyName =
-                    bindData.weightPropertyExpr->ptrCast<PropertyExpression>()->getPropertyName();
-            }
-            GDSUtils::runFrontiersUntilConvergence(context, *gdsComputeState, graph,
+            computeState->initSource(sourceNodeID);
+            GDSUtils::runRecursiveJoinEdgeCompute(context, *computeState, graph,
                 bindData.extendDirection, bindData.upperBound, sharedState->getOutputNodeMaskMap(),
                 propertyName);
-            auto vertexCompute =
-                std::make_unique<RJVertexCompute>(clientContext->getMemoryManager(),
-                    sharedState.get(), rjCompState.outputWriter->copy());
-            auto frontierPair = gdsComputeState->frontierPair.get();
-            auto& candidates = frontierPair->getVertexComputeCandidates();
-            candidates.mergeSparseFrontier(frontierPair->getNextSparseFrontier());
-            if (candidates.enabled()) {
-                GDSUtils::runVertexComputeSparse(candidates, graph, *vertexCompute);
-            } else {
+            auto writer = function->getOutputWriter(context, bindData, *computeState, sourceNodeID,
+                sharedState.get());
+            auto vertexCompute = std::make_unique<RJVertexCompute>(
+                clientContext->getMemoryManager(), sharedState.get(), writer->copy());
+            switch (computeState->frontierPair->getState()) {
+            case GDSDensityState::DENSE: {
```

Comment:
I don't think neither the operator nor `computeState->frontierPair` should be aware of SPARSE and DENSE of internal GDS objects.
1. we should get rid of the switch/case here. All computation should go through the generic interface `GDSUtils::runVertexCompute(context, graph, *vertexCompute);`
2. we should remove the specialized interface for sparse structures `VertexCompute::vertexCompute(common::table_id_t)`, which is not really needed.
3. `VertexCompute` needs expose an interface to indicate if it can run in parallel or not. In the sparse case, we can make sure we only grab one morsel from `FrontierMorselDispatcher`, thus only spawn one thread.